core/node/vm_runner/src/impls/bwip.rs

```diff
@@ -0,0 +1,447 @@
+
+        let blob_url = self.object_store.put(l1_batch_number, &result).await?;
+
+        tracing::info!("Saved VM run data for L1 batch {}", l1_batch_number.0);
+
+        self.pool
+            .connection()
+            .await
+            .unwrap()
+            .proof_generation_dal()
+            .save_vm_runner_artifacts_metadata(l1_batch_number, &blob_url)
+            .await
+            .unwrap();
+
+        Ok(())
+    }
+}
+
+async fn get_updates_manager_witness_input_data(
+    connection: &mut Connection<'_, Core>,
+    updates_manager: Arc<UpdatesManager>,
+) -> VMRunWitnessInputData {
+    let l1_batch_number = updates_manager.l1_batch.number.clone();
+    let finished_batch = updates_manager
+        .l1_batch
+        .finished
+        .clone()
+        .expect(format!("L1 batch {l1_batch_number:?} is not finished").as_str());
+
+    let initial_heap_content = finished_batch.final_bootloader_memory.unwrap(); // might be just empty
+    let default_aa = updates_manager
+        .base_system_contract_hashes()
+        .default_aa
+        .clone();
+    let bootloader = updates_manager
+        .base_system_contract_hashes()
+        .bootloader
+        .clone();
+    let bootloader_code_bytes = connection
+        .factory_deps_dal()
+        .get_sealed_factory_dep(bootloader)
+        .await
+        .expect("Failed fetching bootloader bytecode from DB")
+        .expect("Bootloader bytecode should exist");
+    let bootloader_code = bytes_to_chunks(&bootloader_code_bytes);
+
+    let account_code_hash = h256_to_u256(default_aa);
+    let account_bytecode_bytes = connection
+        .factory_deps_dal()
+        .get_sealed_factory_dep(default_aa)
+        .await
+        .expect("Failed fetching default account bytecode from DB")
+        .expect("Default account bytecode should exist");
+    let account_bytecode = bytes_to_chunks(&account_bytecode_bytes);
+
+    let hashes: HashSet<H256> = finished_batch
+        .final_execution_state
+        .used_contract_hashes
+        .iter()
+        // SMA-1555: remove this hack once updated to the latest version of `zkevm_test_harness`
+        .filter(|&&hash| hash != h256_to_u256(bootloader))
+        .map(|hash| u256_to_h256(*hash))
+        .collect();
+    let mut used_bytecodes = connection
+        .factory_deps_dal()
+        .get_factory_deps(&hashes)
+        .await;
+    if finished_batch
+        .final_execution_state
+        .used_contract_hashes
+        .contains(&account_code_hash)
+    {
+        used_bytecodes.insert(account_code_hash, account_bytecode);
+    }
+
+    let storage_refunds = finished_batch.final_execution_state.storage_refunds;
+    let pubdata_costs = Some(finished_batch.final_execution_state.pubdata_costs);
+
+    VMRunWitnessInputData {
+        l1_batch_number,
+        previous_aux_hash: H256::zero(),
+        previous_meta_hash: H256::zero(),
+        previous_root_hash: H256::zero(),
```

Comment:
It doesn't look right to fill some fields with blank values and set them later. Please try to refactor code to avoid mutations